File: Count distinct elements in every window.cpp
//Problem: Count distinct elements in every window
//geeksforgeeks problem of the day -> 7th January 2026
// C++ CODE
class Solution {
  public:
    vector<int> countDistinct(vector<int> &arr, int k) {
        //CodeGenius
        unordered_map<int,int>mp;
        vector<int>ans;
        
        for(int i=0;i<k;i++)
        mp[arr[i]]++;
        
        ans.push_back(mp.size());
        
        for(int i=k;i<arr.size();i++){
            mp[arr[i]]++;
            mp[arr[i-k]]--;
            
            if(mp[arr[i-k]]==0) mp.erase(arr[i-k]);
            ans.push_back(mp.size());
        }
        return ans;
    }
};

File: Count distinct elements in every window.java
//Problem: Count distinct elements in every window
//geeksforgeeks problem of the day -> 7th January 2026
//JAVA CODE
class Solution {
    ArrayList<Integer> countDistinct(int arr[], int k) {
        int n = arr.length;
        ArrayList<Integer> res = new ArrayList<>();
        Map<Integer, Integer> freq = new HashMap<>();
        for (int i = 0; i < k; i++) {
            freq.put(arr[i], freq.getOrDefault(arr[i], 0) + 1);
        }
        res.add(freq.size());
        for (int i = k; i < n; i++) {
            freq.put(arr[i], freq.getOrDefault(arr[i], 0) + 1);
            freq.put(arr[i - k], freq.get(arr[i - k]) - 1);
            if (freq.get(arr[i - k]) == 0) {
                freq.remove(arr[i - k]);
            }
            res.add(freq.size());
        }
        return res;
    }
}